Initial support spectators

vaut opened the pull request:
These changes add observers to the game.
The definition of losers was completely redesigned and moved to a separate file.
New functionality:
- losers open the entire map. A very useful addition in FFA mode.
- players who cannot play from the very beginning of the game are marked by an observer and the whole map is opened to them.
- a message to the observer about the end of the game.
- punishment for passive play. If a player does not deal damage, does not build structures, does not complete research in 5 minutes, he is considered a loser.

Thanks to these changes to the map with the observer are easier to create. The smallest map with an observer looks like this:
```
level SK-rush2-observer
players 5
type 14
dataset MULTI_CAM_1
game "multiplay / maps / 4c-rush2.gam"
```

If in the game menu add the ability to enter slots without HQ, then the observers will work on all* maps.
*it’s not always always sometimes the empty 7th slot turns out to be scavengers

These changes have not yet been reliably tested.

KJeff01 commented:
You'll need to rebase this since your other PR was merged.

past-due commented:
> These changes have not yet been reliably tested.

@vaut: Have you had a chance to test these changes since? Is this PR in a state it should be reviewed, or were you planning more work / tweaks?

vaut commented:
> > These changes have not yet been reliably tested.
> 
> @vaut: Have you had a chance to test these changes since? Is this PR in a state it should be reviewed, or were you planning more work / tweaks?

The last big change was 12 days ago. I have not conducted large tests in different scenarios. In ordinary team games there are no problems.
> > If in the game menu add the ability to enter slots without HQ, then the observers will work on all* maps.
> > *it’s not always always sometimes the empty 7th slot turns out to be scavengers
> 
> Just to be clear: a spectator is a slot without an associated HQ defined in the map?

HQ does not affect the definition of observers. Players who cannot play at the start become observers.
Criteria for the inability to play:
no factories and no units
or
no factories, units only builders and no reachable oil rigs

So a player who does not have anything at the start becomes an observer. Old maps with an isolated observer base are also processed correctly.


> In what cases does the 7th slot turn out to be scavengers? (i.e. How should we try to detect this? I haven't looked at all the map code in a while.)

I know only about one way to "break" the game:
1. creates a map in Flame for 4 players
2. in the xplayers.lev file we change the number of players by 8
3. The player on the 7th slot begins to manage the scavengers

I did not encounter a problem with maps storing information in json.
[8c-broken_7_slot.zip](https://github.com/Warzone2100/warzone2100/files/4550295/8c-broken_7_slot.zip)

KJeff01 commented on data/mp/multiplay/script/condition.js:
```diff
@@ -0,0 +1,114 @@
+	{
+		if (inOneTeam(playnum, splaynum) && countDroid(DROID_ANY, splaynum) === countDroid(DROID_CONSTRUCT, playnum))
+		{
+			feature["onlyConsttruct"] = true;
```
Replace all instances of "onlyConsttruct" with "onlyConstruct".

KJeff01 commented:
Didn't do a full review so far.

I'd rather you adhere to the style as shown in the introduction in our codingStyle [document](https://github.com/Warzone2100/warzone2100/blob/master/doc/CodingStyle.md)
to better match the rest of the rules.js file. Putting things all on one long line is very hard to read and you'll miss some syntax related issues. I already see a few missing semicolons because of the current style.

I've been using the [jshint](https://jshint.com/) website for linting our scripts. It catches some more things than the LGTM analysis cares to report.

vaut commented:
Thanks for the tool. Fixed all warnings and formatting.

There are a few more dubious places left:
```suggestion
	var oils = enumFeature(ALL_PLAYERS).filter(function(e){if (e.stattype === OIL_RESOURCE) return true; return false;});
```
```suggestion
	if ((alliancesType == ALLIANCES_UNSHARED || alliancesType == ALLIANCES_TEAMS) && (playerData[playnum].team != playerData[splaynum].team))
```

I would be grateful if you tell me how to arrange it more correctly.

```
	setTimer("checkEndConditions", 3000);
	hackNetOn();
	setTimer("gameCompletion", 3000);
```

I create two timers for one entity. I have to use this to bypass the synchronization screen displaying victory.

KJeff01 commented:
```
var oils = enumFeature(ALL_PLAYERS).filter(function(e) {
    if (e.stattype === OIL_RESOURCE) 
    {
        return true;
    }
    return false;
});
```

or

```
var oils = enumFeature(ALL_PLAYERS).filter(function(e) {
    return e.stattype === OIL_RESOURCE;
});
```



```
if ((alliancesType == ALLIANCES_UNSHARED ||
      alliancesType == ALLIANCES_TEAMS) &&
      (playerData[playnum].team != playerData[splaynum].team))
{

}
```

That is what I would do. You don't have to strictly follow the guide, just as long as there isn't any very long lines with potentially multiple lines of code in them.

You shouldn't need two timers. Probably an error somewhere. I'd have to test this later or past-due could.

KJeff01 commented:
Hmm, I tried a [refactor](https://github.com/KJeff01/warzone2100/commits/vautRulesStuff) a few days ago but got stuck on a desync issue. Functions setPower() and removeObject() are dangerous since they affect the state of the game. Feel free to pickup anything in that commit.

vaut commented:
I also ran into this problem.
Code executed after hackNetOn() is not synchronized. Any change in the state of the game will cause a desync.
It seems you can get around this with eventSyncRequest(). But it seems to me that using his bad idea.

I found errors in myself:
- the code for transferring to observers works out every 3 seconds even if the player is already an observer, which leads to a constant change of music.
- In FFA mode, winners are not always correctly determined.
- It also looks like the removal of observer structures occurs in different frames at the host and clients. And it turns out 1 frame desync at the beginning of the game.

At the weekend I’m trying to apply your best practices and fix bugs.

vaut commented:
Good day.
I refactored on KJeff01's ideas. I also conducted a number of tests and fixed the errors found.
I think the changes are ready to merge.
But it would be nice to get a review of more experienced developers.

The attachment has a test card.
0, 1 - cannot play the game from the start.
9 - loses from scavengers.
other valid players.

[10c-Test_condition.zip](https://github.com/Warzone2100/warzone2100/files/4690262/10c-Test_condition.zip)

past-due commented:
Have not had a chance to test this yet.

A few thoughts:
- We should probably prevent spectators from texting with normal players, since they could reveal information about the map / other players. (Spectators can text with spectators, though.)
- We should incorporate a way of handling and displaying this in the game / lobby setup UI.
   - Which makes me think it might be better to port this `canPlay()` logic to the C++ side and expose a JS API for determining whether a player # is a spectator or a player.

berserkingyadis commented:
In some competitive RTS games there are "Referee" type Spectators who can write with the Players, maybe we can adapt it.

KJeff01 commented on data/mp/multiplay/skirmish/rules.js:
```diff
@@ -396,82 +409,85 @@ function eventGameInit()
-		removeTimer("checkEndConditions");
+		if (win)
+		{
+			if (playerData[playnum].condition == CONDITION.victory.winer)
+				{
+					continue;
+				}
+			else {
+				playerData[playnum].condition = CONDITION.victory.winer;
```
winer -> winner

vaut commented:
Found unacceptable performance degradation in NTW 5x5 games
Already have thoughts on how this can be fixed.

vaut commented:
Again, I changed most of the logic for determining the status of players. There were serious errors in the logic.
Blinded all the commits with the correct syntax and fix broken logic.
I hope the code has become more understandable and someone will have time for a full review with a full check of the changes.

Quick tests did not reveal the problems. Unsubscribe after a more complete check.

PS added a small report at the end of the game as a bonus.

vaut commented:
The code definitely works well in 1x1 mode with an observer. These changes have been used in automated games for several months now. Unfortunately, I cannot host, as my IP is occupied by them.

I hope for your help in testing and reviewing.
Potentially dangerous places:
- desync when removing buildings when transferring to observers.
Sometimes he skipped in MP games, but later the game was synchronized.
- incorrect work after reporting a victory or defeat.

vaut commented:
[SPEC_d5d0112.zip](https://github.com/Warzone2100/warzone2100/files/5402591/SPEC_d5d0112.zip)

vaut commented:
Rebase of changes to the new format for storing scripts required deleting the entire history of changes.
But I don't think she was important.

The capabilities of the new JS engine require the mod to be finalized.
Not for a long time I mark it WIP

vaut commented:
This change not only adds observers, but also correctly opens the entire map to the loser.
Also, the inactivity timer eliminates the problem of hiding in the pool at the border of other non-sports activities.
I ask you to review this patch again and apply it if the code meets the code standards.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
+			count += countDroid(DROID_ANY, splaynum) - countDroid(DROID_CONSTRUCT, splaynum);
+		}
+	}
+	if (count === 0) {return true;}
+	else {return false;}
```
Indent these control statements.

if ()
{
  xxx
}

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
-	for (var playnum = 0; playnum < maxPlayers; playnum++)
+function canPlay(playnum)
+{
+/*
+	let feature = {
+		factory: hasFactory(playnum),
+		droid: hasDroid(playnum),
+		onlyConstruct: hasOnlyConstructor(playnum),
+		oilReach: canReachOil(playnum),
+		activeGame: activeGame(playnum)
+	};
+	debug(playerData[playnum].name + JSON.stringify(feature));
+*/
```
You can remove this.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
-			droids = countDroid(DROID_ANY, playnum);
-			if (droids > 0 || factories > 0)
+			if (playnum != splaynum &&
+			playerData[playnum].usertype == USERTYPE.player.fighter &&
+			playerData[splaynum].usertype == USERTYPE.player.fighter &&
+			!inOneTeam (playnum, splaynum))
```
This looks weird (at least in GitHub). Maybe add tabs here if applicable.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
 	}
+	if (unPlayability)
+	{
+//		debug(_("There are no opponents in the game."));
+		console(_("There are no opponents in the game"));
+//		gameOverMessage(true);
```
Consider if these comments are needed anymore.

vaut replied:
remove comments.

past-due commented:
I'm working on a player slot refactor that will handle this in the core, and I believe spectator slot support in the core is the best approach. It lets us do a number of things (like not letting spectator slot networking delay actual players), which can't be accomplished with just script changes.

I've been able to port most of the logic from these changes to the WZ core, and when I have a chance I'll rebase it and open a PR.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
+{
+	hackNetOff();
+	setPower(0, playnum);
+	let spotter = {
+		X: mapWidth/2,
+		Y: mapHeight/2,
+		radius: Math.sqrt(mapWidth*mapWidth + mapHeight*mapHeight)/2*128
+	};
+	addSpotter(spotter.X, spotter.Y, playnum, spotter.radius, 0, 0);
```
Could decapitalize the X and Y here.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
+		}
+		if (playerData[selectedPlayer].usertype == USERTYPE.spectator)
+		{
+//			gameOverMessage(true);
+			console(_("the battle is over, you can leave the room"));
+//			debug("the battle is over, you can leave the room");
```
Comments needed?

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
+	for (let playnum = 0; playnum < maxPlayers; playnum++)
+	{
+		playerData[playnum].lastActivity = gameTime;
+			//we consider observers to players who cannot play from the beginning of the game
```
Extra tab.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
+	if (playerData[selectedPlayer].lastActivity + IDLETIME/2 < gameTime)
+	{
+		console(_("Playing passively will lead to defeat. Actions that are considered: - unit building - research completion - construction of base structures (factories, power plants, laboratories, modules and oil derricks) - dealing damage"));
+	//		debug (getMissionTime());
```
Extra tab. Maybe you don't need this debug anymore.

KJeff01 commented on data/mp/multiplay/script/rules/endconditions.js:
```diff
@@ -1,62 +1,414 @@
+	let structs = [FACTORY, CYBORG_FACTORY, VTOL_FACTORY];
+
+	for (let splaynum = 0; splaynum < maxPlayers; splaynum++)
+		{
```
This brace is too far to the right. Remove a tab.

KJeff01 commented:
It looks fine too me. I only added some suggestions for code clarity.

Does this cause any desyncs? You've probably tested this for a long time already so probably not, huh.

KJeff01 commented:
> I'm working on a player slot refactor that will handle this in the core, and I believe spectator slot support in the core is the best approach. It lets us do a number of things (like not letting spectator slot networking delay actual players, adding proper replay support, not requiring map edits to have spectators, etc), which can't be accomplished with just script changes.
> 
> I've been able to port most of the logic from these changes to the WZ core, and when I have a chance I'll rebase it and open a PR.

And after I just reviewed this...

vaut commented:
Thanks a lot for the review. Removed unnecessary comments.
Fixed the style.

Once the inline observers are added, there is not much code to change:
- remove the definition of observers at the start
- replace the toSpectator() function if the API is expanded.

The change in the state of the game is calculated for all players at the same time. To prevent a large number of concurrent events from causing a temporary desync, they are wrapped in calls to hackNetOff(); hackNetOn();

PS GitHub is in no hurry to push the second commit with correcting comments to the pull request.

vaut commented:
[spectator_4.0.1.zip](https://github.com/Warzone2100/warzone2100/files/6625335/spectator_4.0.1.zip)

vaut commented:
Use these changes:
#1957
Perhaps in the future there will be a mod taking into account the specifics of observers based on a new pull request
